### PR TITLE
Add Undo/Redo to dietline emacs mode

### LINKED ISSUE #572: Ctrl-_/Ctrl-/ and Meta-_/Meta-/ - Undo and Redo
**Is your feature request related to a problem? Please describe.**
Sometimes you accidentally delete part of the command you are writing, to aid the clumsy user it would be nice to have an undo and redo keyboard shortcuts.

**Describe the solution you'd like**
Following the emacs/readline shortcuts:
- `Ctrl-_` or `Ctrl-/` for undoing
- `Meta-_` or `Meta-/` for redoing

Pressing multiple times undo/redo further.

**Describe alternatives you've considered**
💔 

**Additional context**
Truth is that readline doesn't seem to have a redo, and vanilla emacs has a "strange" redo: `Ctrl-_` or `Ctrl-/` for undoing, `Ctrl-g` for reversing the undo-action (not yet redid anything and only works right after an undo), now, after `Ctrl-g`, `Ctrl-_` or `Ctrl-/` perform redo until another `Ctrl-g` is pressed to reverse the meaning of `Ctrl-_` or `Ctrl-/`.

Some of the modern popular emacs packages/frameworks have `Ctrl-_` or `Ctrl-/` mapped as undo and `Meta-_` or `Meta-/` as redo to avoid the madness.

### PR DESCRIPTION
**Pull request by kui8shi**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Bash and Emacs have the functionality of Undo and Redo. I added code corresponding to those ones into dietline which is used in RzShell. They can be triggered by Ctrl-/ Ctrl-_ for Undo, and Meta-/ Meta-_ for Redo. For that RzLine struct has had a new vector "undo_vec" and an index "undo_cursor". When executing an Undo or Redo action, the cursor moves back and forth while vecotor entries restore the command buffer condition.
As far as I tested manually it works but I am going to add unit tests. Also, I am willing to add Doxygen comments to new functions if necessary.

bonus: I fixed a bug at unix_word_rubout since it cliped too much strings and hindered correct addition of undo_entry.

**Test plan**

To do: add unit tests.

**Closing issues**

closes #572

### CODE REVIEW

File: librz/include/rz_cons.h
Comment (ret2libc) [inline librz/include/rz_cons.h]: are these values really important? If so, maybe add a doxygen documentation to the whole `enum` and explain a bit more.

File: librz/include/rz_cons.h
Comment (ret2libc) [inline librz/include/rz_cons.h]: add doxygen doc

File: librz/cons/dietline.c
Comment (ret2libc) [inline librz/cons/dietline.c]: call this something else, not starting with `__`. e.g. `line_do`

File: librz/cons/dietline.c
Comment (ret2libc) [inline librz/cons/dietline.c]: use doxygen doc

File: librz/cons/dietline.c
Comment (ret2libc) [inline librz/cons/dietline.c]: avoid names starting with `__`

File: librz/cons/dietline.c
Comment (ret2libc) [inline librz/cons/dietline.c]:
```suggestion
	RZ_FREE(e->text);
```

File: librz/cons/dietline.c
Comment (ret2libc) [inline librz/cons/dietline.c]:
```suggestion
	return !!I.undo_vec;
```

File: librz/cons/dietline.c
Comment (ret2libc) [inline librz/cons/dietline.c]:
```suggestion
static bool concat_undo_entry(char *diff) {
```

File: librz/include/rz_cons.h
Comment (ret2libc) [inline librz/include/rz_cons.h]: it is not clear when/how we should use these... why are they necessary? Please add doc for these 4 functions and also some doc explaining the meaning and usage of the RzLineUndoType.

File: librz/cons/dietline.c
Comment (wargio) [reply librz/cons/dietline.c]: also if you are creating static functions, avoid to set the `inline` attribute..

File: librz/include/rz_cons.h
Comment (wargio) [reply librz/include/rz_cons.h]: these should be `RZ_IPI` and these needs to be moved under `core_private.h`

File: librz/include/rz_cons.h
Comment (kui8shi) [reply librz/include/rz_cons.h]: In some way, I have to tell the undo system whether a single action is to delete text or insert a text, and how many actions does a group of an undo or redo have (especially in vi mode, the number of contained actions will increase). The reason that the enum has a value was just to make it easy to reverse the type of an action.
However, I now am thinking I should do this with another option, which is to distinguish text to be deleted (in the sense of undo) and text to be inserted, and label each entry with an identifer without having special entries at the beginning and end. This might be more straightforward and readable.

File: librz/include/rz_cons.h
Comment (kui8shi) [reply librz/include/rz_cons.h]: I will remove these functions.

File: librz/cons/dietline.c
Comment (XVilka) [inline librz/cons/dietline.c]: If you flip the condition, you could use early return pattern and avoid nesting.

File: librz/cons/dietline.c
Comment (XVilka) [inline librz/cons/dietline.c]: Same here

File: librz/cons/dietline.c
Comment (XVilka) [inline librz/cons/dietline.c]: We use `\brief` not `@brief` syntax

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/include/rz_cons.h
Comment (XVilka) [reply librz/include/rz_cons.h]: You also forgot this

File: librz/cons/dietline.c
Comment (XVilka) [inline librz/cons/dietline.c]: Return type should be `bool`

File: librz/include/rz_cons.h
Comment (XVilka) [inline librz/include/rz_cons.h]: `bool` as the return, `void` for the `_free()`

**Review by XVilka — CHANGES_REQUESTED** [review]
Two minor things, apart from that look good from the code point of view. Didn't test yet. Please also rebase on top of the latest `dev`.

File: librz/include/rz_cons.h
Comment (wargio) [reply librz/include/rz_cons.h]: i'm not 100% sure if these should be RZ_API, but i guess these could be ok now.

Comment (kui8shi) [reply librz/include/rz_cons.h]: I can minimize the surface, should I do that?

File: librz/cons/dietline.c
Comment (XVilka) [inline librz/cons/dietline.c]: also `bool`.

File: librz/cons/dietline.c
Comment (XVilka) [inline librz/cons/dietline.c]: Please also make this one `bool` as well, while you are on it.

**Review by XVilka — APPROVED** [review]
Good job, LGTM once you address these two more nitpicks

**Comment by kui8shi** [comment]
Excuse me for the additional commit.

**Comment by XVilka** [comment]
Please rebase on top of the `dev`, also squash all commits into one (or just few, grouped semantically), there are seem to be some conflicts

**Review by XVilka — APPROVED** [review]
Good job! Now it's good as is. Will merge after the second approval

File: test/unit/test_cons.c
Comment (ret2libc) [inline test/unit/test_cons.c]: I think this is not necessary. Did you just copy this from `test_line_onecompletion` or was it actually required for the test to work?

File: test/unit/test_cons.c
Comment (ret2libc) [inline test/unit/test_cons.c:336]: is the "undo" the `\x1f`? Is the `\n` after `\x1f` required?

File: test/unit/test_cons.c
Comment (ret2libc) [inline test/unit/test_cons.c]: same here, is the `\n` required?

File: librz/include/rz_cons.h
Comment (ret2libc) [inline librz/include/rz_cons.h]: is this really required? I think it can be moved in `line.c` and just be static.
There is no `rz_line_undo_new` so I don't expect a user of RzCons to call `rz_line_undo_free` either.

File: librz/include/rz_cons.h
Comment (ret2libc) [inline librz/include/rz_cons.h]: I think this whole structure is not even necessary here at this point and it could be hidden inside line.c or dietline.c, because the users of RzCons don't require any direct access to it, right?

**Review by ret2libc — COMMENTED** [review]
I will review it more and test it later today and give a final review.

File: test/unit/test_cons.c
Comment (kui8shi) [reply test/unit/test_cons.c:336]: undo is '\x1f' and redo is '\x1b\x3f'. For '\n', it was just needed for return from readline func. I will do the tests in more direct way.

File: librz/include/rz_cons.h
Comment (kui8shi) [reply librz/include/rz_cons.h]: exactly right. I will hide it.

File: test/unit/test_cons.c
Comment (kui8shi) [reply test/unit/test_cons.c:336]: It appears that I have to use '\n'.

File: librz/cons/dietline.c
Comment (kui8shi) [inline librz/cons/dietline.c]: when the additional test was added, I realized that I have to change the way to process group operations. so I added undo_continue which indicates continuous undo entries, instead of an argument of undo_add_entry().

File: librz/cons/dietline.c
Comment (kui8shi) [inline librz/cons/dietline.c:117]: made them static. I will remove comments if they are noisy.

**Comment by XVilka** [comment]
```
   Creating library binrz/rz-hash/rz-hash.lib and object binrz/rz-hash/rz-hash.exp
rz_cons.lib(line.c.obj) : error LNK2019: unresolved external symbol rz_line_undo_add_entry referenced in function rz_line_readline_cb
binrz/rz-hash/rz-hash.exe : fatal error LNK1120: 1 unresolved externals
[1686/1946] Linking target binrz/rz-run/rz-run.exe
FAILED: binrz/rz-run/rz-run.exe 
"link"  /MACHINE:x64 /OUT:binrz/rz-run/rz-run.exe binrz/rz-run/rz-run.exe.p/rz-run.c.obj "/nologo" "/release" "/nologo" "/OPT:REF" "librz\main\rz_main.lib" "librz\util\rz_util.lib" "subprojects\zlib-1.2.13\libz.a" "subprojects\xz-5.4.1\src\liblzma\liblzma.a" "librz\demangler\rz_demangler.lib" "subprojects\libdemangle\libdemangle.a" "librz\magic\rz_magic.lib" "librz\socket\rz_socket.lib" "librz\flag\rz_flag.lib" "librz\cons\rz_cons.lib" "librz\hash\rz_hash.lib" "subprojects\xxhash\libxxhash.a" "subprojects\blake3\libblake3.a" "librz\crypto\rz_crypto.lib" "subprojects\nettle\libnettle-aes.a" "librz\il\rz_il.lib" "librz\reg\rz_reg.lib" "librz\io\rz_io.lib" "subprojects\rzgdb\librzgdb.a" "subprojects\rzwinkd\librzwinkd.a" "subprojects\libzip-1.9.2\liblibzip.a" "subprojects\rzar\librzar.a" "subprojects\rzqnx\librzqnx.a" "subprojects\rzw32dbg_wrap\librzw32dbg_wrap.a" "librz\bp\rz_bp.lib" "librz\syscall\rz_syscall.lib" "librz\parse\rz_parse.lib" "librz\asm\rz_asm.lib" "librz\config\rz_config.lib" "librz\bin\rz_bin.lib" "librz\type\rz_type.lib" "subprojects\tree-sitter-c\libtree-sitter-c.a" "subprojects\tree-sitter-0.20.7\libtree_sitter.a" "subprojects\lz4-1.9.4\liblz4.a" "subprojects\yxml\libyxml.a" "subprojects\libmspack\liblibmspack.a" "subprojects\capstone-bundled\libcapstone.a" "subprojects\rzspp\librzspp.a" "librz\egg\rz_egg.lib" "subprojects\mpc\libmpc.a" "librz\search\rz_search.lib" "librz\analysis\rz_analysis.lib" "librz\diff\rz_diff.lib" "librz\debug\rz_debug.lib" "librz\sign\rz_sign.lib" "librz\core\rz_core.lib" "librz\lang\rz_lang.lib" "subprojects\rizin-shell-parser\libshell_parser.a" "ws2_32.lib" "wininet.lib" "psapi.lib" "ws2_32.lib" "wininet.lib" "psapi.lib" "advapi32.lib" "bcrypt.lib" "ws2_32.lib" "wininet.lib" "psapi.lib" "advapi32.lib" "bcrypt.lib" "ws2_32.lib" "wininet.lib" "psapi.lib" "ws2_32.lib" "wininet.lib" "psapi.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
   Creating library binrz/rz-run/rz-run.lib and object binrz/rz-run/rz-run.exp
rz_cons.lib(line.c.obj) : error LNK2019: unresolved external symbol rz_line_undo_add_entry referenced in function rz_line_readline_cb
binrz/rz-run/rz-run.exe : fatal error LNK1120: 1 unresolved externals
[1687/1946] Compiling C object binrz/rz-asm/rz-asm.exe.p/rz-asm.c.obj
ninja: build stopped: subcommand failed.
```

File: librz/cons/dietline.c
Comment (XVilka) [reply librz/cons/dietline.c]: Please avoid globals and use context, like `RzCons`, for example.

Comment (XVilka) [reply librz/cons/dietline.c]: See https://github.com/rizinorg/rizin/issues/276 for more information about efforts of removing global state completely.

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Comment by ret2libc** [comment]
I have troubles using the "redo"/Meta-_ actually. I'm on MacOS iTerm, but also on Terminal it doesn't seem to work. Is anyone else using MacOS?

File: librz/cons/dietline.c
Comment (ret2libc) [inline librz/cons/dietline.c]: I would just ignore this case actually and just do a free + new anyway. I don't think we save too much by doing this and at least we are sure we don't miss any case (e.g. variables that might have a different value than expected after a reset, like `undo_continue`)

**Comment by XVilka** [comment]
> I have troubles using the "redo"/Meta-_ actually. I'm on MacOS iTerm, but also on Terminal it doesn't seem to work. Is anyone else using MacOS?

I think these keys trigger system-wide actions and can't be passed to the terminal (it resizes the iTerm window in my case)

**Review by ret2libc — APPROVED** [review]
Code looks good overall! I could not test the redo though.

File: librz/cons/dietline.c
Comment (kui8shi) [reply librz/cons/dietline.c]: So, to confirm "undo_continue" is false is enough?  Although I am happy to delete this case.

**Comment by kui8shi** [comment]
I feel changing the key bindings is an option (e.g. in Nano, undo is '^[u' and redo is '^[e'), like undo = '^[u', and redo = '^[r'.

File: librz/cons/dietline.c
Comment (ret2libc) [reply librz/cons/dietline.c]: I think you can delete this case altogether.

**Comment by ret2libc** [comment]
Do you have any idea why the redo would not work with Ctrl?

**Comment by kui8shi** [comment]
First, only few spaces are left for ctrl+*. Secondly, as far as I know, ctrl+shift+/ (ctrl+?) does not work because it has no keycode.

**Review by XVilka — APPROVED** [review]
I think it's good as is, we can modify bindings later if necessary. Thank you for your patience!

**Pull request merged**
